### ISSUE TITLE
Change encoding format re: alpha channel

### ISSUE THREAD

**Issue by aldanor** [nigeltao/qoi2-bikeshed#12]
https://github.com/phoboslab/qoi/issues/38 and implementation in https://github.com/phoboslab/qoi/pull/41

**Comment by oscardssmith** [comment]
Here would be my proposed encoding format. The key difference is that it sacrifices some of the cache and RLE codes to get a full 5 bits for each color into the 16 bit diff opcode.

0: QGI_DIFF_16
* color mode: use 5 bits for rgb delta
* alpha mode: use 4 bits for rgb and 3-bits delta for alpha

10: QGI_DIFF_8
* 2 bits delta for rgb

1100: QGI_INDEX
* use 4-bit index where each index references a bucket of 8 possible entries. This lets you use avx instructions to efficiently find the correct match without significant increases in encoding time.

1101: QGI_RUN_8
* 4-bit rep count, chained so rle commands get turned into 1

1110: QGI_DIFF_24
* color mode: 6 bits blue, 7 bits red and green delta
* alpha mode: 5 bits for r,g,b and a

1111 full with mask, same for alpha and color mode
exactly the same as qoi: 4 bit channel mask, then individual channel bytes if present
except that if mask is 0, the decoder should flip between modes (color<->alpha) and continue processing (no pixel is output in this case)

**Comment by nigeltao** [comment]
> use 4-bit index where each index references a bucket of 8 possible entries.

How does that work? Which of the 8 do you pick when decoding?

It might be clearer if you can share some prototype code.

**Comment by oscardssmith** [comment]
good point. that doesn't work. (I guess that's the result of trying to think early in the morning). the one potentially good part of this idea is giving run length a longer tag, which would help if we have another 4 bit tag that we want to use.

**Comment by oscardssmith** [comment]
Here's my updated version of the proposal now that my brain is working again. This reverts the change for `QOI_INDEX` and gives an extra bit to `QOI_DIFF_24`

00: QOI_DIFF_8
* 2 bits delta for rgb

01: QOI_DIFF_16
* color mode: 5 bits delta for rg delta, 4 bits for b
* alpha mode: 4 bits delta for rgba

10: QOI_INDEX
* use 6-bit index

110: QOI_DIFF_24
* color mode: 7 bits delta for rgb
* alpha mode: 5 bits delta for 5gb, 6 bits for a

1110: QOI_RUN_8
* 4-bit rep count, chained so rle commands get turned into 1

1111: QOI_COLOR

**Comment by oscardssmith** [comment]
The other alternative would be to add a `QOI_INDEX_DIFF` tag that would use a 6 bit index and 2 bit RGB delta. This could be encoded relatively efficiently by storing 2 indexes: one indexed as currently by QOI_HASH(RGBA), and the other indexed by only the top 6 bits of the pixel values.
